### ISSUE TITLE
How to completely clear a Map?

### ISSUE THREAD

**Issue by archseer** [CosmWasm/cw-plus#447]
I noticed there's no `.clear()` method. Tried doing something like this:

```rust
pub const FOO: Map<&Addr, ()> = Map::new("foo");

for key in FOO.keys(deps.storage, None, None, Order::Ascending) {
    FOO.key(&key).remove(deps.storage);
}
```

But this fails since keys return `&Vec<u8>` instead of `&Addr`.

**Comment by archseer** [comment]
Looks like `deps.storage.remove(&key)` might be what I need.

**Comment by hashedone** [comment]
Technically `deps.storage.remove` removes raw key (without anything like namespacing), so to use it you need to properly build a key (and honestly looking at code I am not 100% sure if keys returned by `Map::keys` are already prefixed, my bet is: no). The problem is actually that keys are not deserialized when returned from map in interfaces, it is related to #198.

Basically what I am pretty sure about (but it is to be verified - @ethanfrey?) is that the `&Vec<u8>` you are getting are just bytes of `Addr`, so to get your `Addr` back you just need to do: `Addr::unchecked(str::from_utf8(&key).unwrap())` - obviously handling unwrap in some reasonable way (probably just `?` to your custom err type). To double check if I am right here, just print your addr after conversion, like:
```rust
let key = dbg!(Addr::unchecked(str::from_utf8(&key).unwrap()));
```
And run some failing test which calls this function (failing one as passing tests hides their output by default).

**Comment by ethanfrey** [comment]
The following should work:

```rust
pub const FOO: Map<&Addr, ()> = Map::new("foo");

for raw_key in FOO.keys(deps.storage, None, None, Order::Ascending) {
    // we need to update the type of key first...
    let key = deps.api.addr_validate(String::from_utf8(raw_key)?)?;
    FOO.remove(deps.storage, &key);
}
```

We are working on automatically deserialising keys into the proper type in https://github.com/CosmWasm/cw-plus/pull/432
so you get Addr rather than Vec<u8>. When that is all finished and merged into cw-plus 0.10, you will be able to do something like:

```rust
pub const FOO: Map<&Addr, ()> = Map::new("foo");

for key in FOO.keys(deps.storage, None, None, Order::Ascending) {
    FOO.remove(deps.storage, &key);
}
```

I don't think adding a `clear` method is a good idea, as it will be very expensive. Basically an unbounded cost hidden in a function. At least doing a range without limit should make one realise what is going on.

**Comment by ethanfrey** [comment]
This is done now. (cw-plus helpers)
